content/posts/2019-03-job-search-2-parsing.md

```diff
@@ -0,0 +1,118 @@
+
+Onwards, let's talk about the steps we took to build our parser!
+
+## Defining the Grammar
+
+We started with an idea of the language we wanted to implement. This is referred to as the grammar. It's common to use [Extended Backus-Naur Form (EBNF)](https://en.m.wikipedia.org/wiki/Extended_Backus–Naur_form)  which is a syntax used to desribe the individual components of the grammar. Here's a snippet of Jobs Query Language (JQL) in EBNF:
+
+```
+<and> ::= 'and' | '&&';
+<or> ::= 'or' | '||';
+<not> ::= 'not' | '-';
+<quote> ::= '"';
+<colon> ::= ':';
+<lparen> ::= '(';
+<rparen> ::= ')';
+<parens> ::= <lparen> | <rparen>;
+<lbracket> ::= '[';
+<rbracket> ::= ']';
+<brackets> ::= <lbracket> | <rbracket>;
+<string> ::= { <any_character> - (<brackets> | <parens> | <quote>) };
+<quoted_string> ::= <quote> { <any_character> - <quote> } <quote>;
+<tag> ::= { <letter> | <number> | '-' | '.' | '#' | '+' | '*' };
+<modifier> ::= { <letter> | '-' } <colon>;
+<range> ::= '.', '.';
+<unit> ::= <letter>, <letter>, <letter> | <letter>, <letter> | <letter>;
+<literal> ::= <number> | <string> | <quoted_string>;
+<term> ::= 
+    <modifier>, <colon>, <literal> |
+    <lbracket>, <tag>, <rbracket> | 
+    <string> | 
+    <quoted_string>;
+    
+<expression> ::= 
+    <lparen>, <expression>, <rparen> | 
+    <not>, <expression> |
+    <expression>, <and>, <expression> | 
+    <expression>, <or>, <expression> |
+    <term>;
+```
+
+This supports a wide variety of inputs for search:
+
+- **Simple Text**: `hello world`
+- **Quoted Text**: `"hello world"`
+- **Tags**: `[asp.net]`
+- **Modifiers**: `salary:10000` or `salary:10000USD`
+- **Ranges**: `salary:10000..20000` or `salary:..50000` or `salary:10000..`
+- **Expressions**: `[asp.net] or "hello world"` or `[c#] and not [java] and salary:10000..20000`
+- **Complex Expressions**: `([asp.net] or "hello world") and (([c#] and not [java]) or salary:10000..)`
+
+Of particular importance is the definition of `<expression>`; this is what allows for the nesting of different parts of the grammar via recursion.
+
+## Building the Parser
+ 
+Once we've defined a grammar, our next step is to break down each rule into a set of mini-parsers. In our case we'd like to take an arbitrary string input and return something representing the parsed form of it. A common way of representing the parsed form is as an [abstract syntax tree (AST)](https://en.wikipedia.org/wiki/Abstract_syntax_tree).
+
+JQL is represented using an AST that uses an abstract base class called `JqlNode`. We have implementations that reflect the structure of the grammar. E.g. it consists of a `QueryNode` representing a query which consists of things like `LiteralNode` to represent text/numbers/bools and `ModifierNode` to handle modifiers like `remote:true`.
```

Comment:
Suggest providing a concrete example here (eg https://www.benjamin.pizza/posts/2017-11-13-recursion-without-recursion.html#asts-and-operations)